File: csharp/Bingo.Test/BingoTestBDD.cs
namespace Bingo.Test;

public class BingoTestBdd
{

    BingoBoard board;

    [Fact]
    void AnNewlyCreatedBoardIsNotInitialized()
    {
        GivenBingoBoardOfSize(1, 1);
        ThenBoardIsNotInitialized();
    }

    [Fact]
    void WhenAllFieldsAreSetTheBoarIsInitialized()
    {
        var anyValue = "42";
        board = new BingoBoard(1, 1);
        board.DefineCell(0, 0, anyValue);
        board.IsInitialized().Should().BeTrue();
    }

    [Fact]
    void WhenAllFieldsOnRectangularBoardAreSetItIsInitialized()
    {
        var one = "one, two, three";
        var two = "Bingo cells can contain any text";
        GivenBingoBoardOfSize(1, 2);
        WhenCellIsDefined(0, 0, one);
        WhenCellIsDefined(0, 1, two);
        ThenBoardIsInitialized();
    }

    [Fact]
    void ADefinedCellCantBeRedefinedEvenIfItsTheSameValue()
    {
        var anyValue = "42";
        GivenBingoBoardOfSize(1, 1);
        WhenCellIsDefined(0, 0, anyValue);
        Action secondCall = () => WhenCellIsDefined(0, 0, anyValue);
        secondCall.Should()
            .Throw<InvalidOperationException>()
            .Where(e => e.Message.Contains("already defined"));
    }

    [Fact]
    void DuplicateCellsAreNotAllowed()
    {
        var anyValue = "42";
        GivenBingoBoardOfSize(2, 2);
        WhenCellIsDefined(0, 1, anyValue);
        Action secondCall = () => WhenCellIsDefined(1, 0, anyValue);
        secondCall.Should()
            .Throw<InvalidOperationException>()
            .WithMessage(anyValue + " already present at 0,1");
    }

    [Fact]
    void ANonInitializedBoardCannotBeMarked()
    {
        GivenBingoBoardOfSize(1, 1);
        Action markCell = () => WhenCellIsMarked(0, 0);
        markCell.Should()
            .Throw<InvalidOperationException>()
            .Where(e => e.Message.EndsWith("not initialized"));
    }

    [Fact]
    void WhenAllCellGetsMarkedItIsMarked()
    {
        var anyValue = "42";
        GivenBingoBoardOfSize(1, 1);
        WhenCellIsDefined(0, 0, anyValue);
        WhenCellIsMarked(0, 0);
        ThenCellIsMarked(0, 0);
    }

    private void GivenBingoBoardOfSize(int width, int height) {
        board = new BingoBoard(width, height);
    }

    private void WhenCellIsDefined(int x, int y, String value) {
        board.DefineCell(x, y, value);
    }

    private void WhenCellIsMarked(int x, int y) {
        board.MarkCell(x, y);
    }

    private void ThenBoardIsNotInitialized()
    {
        BoardInitializeState().Should().BeFalse();
    }

    private void ThenBoardIsInitialized() {
        BoardInitializeState().Should().BeTrue();
    }

    private bool BoardInitializeState() {
        return board.IsInitialized();
    }

    private void ThenCellIsMarked(int x, int y) {
        board.IsMarked(x, y).Should().BeTrue();
    }
}

File: php/tests/BingoTestBDD.php
<?php

namespace KataTest;

use Kata\BingoBoard;
use PHPUnit\Framework\TestCase;

class BingoTestBDD extends TestCase
{
    private BingoBoard $board;

    public function testANewlyCreatedBoardIsNotInitialized(): void
    {
        $this->givenBingoBoardOfSize(1, 1);
        $this->thenBoardIsNotInitialized();
    }

    public function testWhenAllFieldsAreSetTheBoardIsInitialized(): void
    {
        $anyValue = "42";
        $this->givenBingoBoardOfSize(1, 1);
        $this->whenCellIsDefined(0, 0, $anyValue);
        $this->thenBoardIsInitialized();
    }

    public function testWhenAllFieldsOnRectangularBoardAreSetItIsInitialized(): void
    {
        $one = "one, two, three";
        $two = "Bingo cells can contain any text";
        $this->givenBingoBoardOfSize(1, 2);
        $this->whenCellIsDefined(0, 0, $one);
        $this->whenCellIsDefined(0, 1, $two);
        $this->thenBoardIsInitialized();
    }

    public function testADefinedCellCantBeRedefinedEvenIfItsTheSameValue(): void
    {
        $anyValue = "42";
        $this->givenBingoBoardOfSize(1, 1);
        $this->whenCellIsDefined(0, 0, $anyValue);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/already defined/');
        $this->whenCellIsDefined(0, 0, $anyValue);
    }

    public function testDuplicateCellsAreNotAllowed(): void
    {
        $anyValue = "42";
        $this->givenBingoBoardOfSize(2, 2);
        $this->whenCellIsDefined(0, 1, $anyValue);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/' . preg_quote($anyValue . " already present at 0,1") . '/');
        $this->whenCellIsDefined(1, 0, $anyValue);
    }

    public function testANonInitializedBoardCannotBeMarked(): void
    {
        $this->givenBingoBoardOfSize(1, 1);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/not initialized/');
        $this->whenCellIsMarked(0, 0);
    }

    public function testWhenAllCellGetsMarkedItIsMarked(): void
    {
        $anyValue = "42";
        $this->givenBingoBoardOfSize(1, 1);
        $this->whenCellIsDefined(0, 0, $anyValue);
        $this->whenCellIsMarked(0, 0);
        $this->thenCellIsMarked(0, 0);
    }

    private function givenBingoBoardOfSize(int $width, int $height): void
    {
        $this->board = new BingoBoard($width, $height);
    }

    private function whenCellIsDefined(int $x, int $y, string $value): void
    {
        $this->board->defineCell($x, $y, $value);
    }

    private function whenCellIsMarked(int $x, int $y): void
    {
        $this->board->markCell($x, $y);
    }

    private function thenBoardIsNotInitialized(): void
    {
        $this->assertFalse($this->boardInitializeState());
    }

    private function thenBoardIsInitialized(): void
    {
        $this->assertTrue($this->boardInitializeState());
    }

    private function boardInitializeState(): bool
    {
        return $this->board->isInitialized();
    }

    private function thenCellIsMarked(int $x, int $y): void
    {
        $this->assertTrue($this->board->isMarked($x, $y));
    }
}
